source/portable/BufferManagement/BufferAllocation.c

```diff
@@ -0,0 +1,647 @@
+                pucEthernetBuffer += ipBUFFER_PADDING;
+            }
+        }
+
+        return pucEthernetBuffer;
+    }
+
+#endif /* if ipconfigIS_DISABLED( ipconfigBUFFER_ALLOC_STATIC ) */
+
+/*-----------------------------------------------------------*/
+
+#if ( ipconfigIS_DISABLED( ipconfigBUFFER_ALLOC_STATIC ) )
+
+    static void vReleaseNetworkBuffer( const uint8_t * const pucEthernetBuffer )
+    {
+        /* There is space before the Ethernet buffer in which a pointer to the
+         * network buffer that references this Ethernet buffer is stored.  Remove the
+         * space before freeing the buffer. */
+        if( pucEthernetBuffer != NULL )
+        {
+            const uint8_t * pucEthernetBufferCopy = pucEthernetBuffer - ipBUFFER_PADDING;
+            vPortFree( ( void * ) pucEthernetBufferCopy );
+        }
+    }
+
+#endif /* if ipconfigIS_DISABLED( ipconfigBUFFER_ALLOC_STATIC ) */
+
+/*-----------------------------------------------------------*/
+
+#if ipconfigIS_DISABLED( ipconfigBUFFER_ALLOC_STATIC )
+
+    NetworkBufferDescriptor_t * pxResizeNetworkBufferWithDescriptor( NetworkBufferDescriptor_t * pxNetworkBuffer,
+                                                                     size_t xNewSizeBytes )
+    {
+        NetworkBufferDescriptor_t * pxNetworkBufferCopy = pxNetworkBuffer;
+        size_t uxSizeBytes = xNewSizeBytes;
+
+        if( baADD_WILL_OVERFLOW( uxSizeBytes, ipBUFFER_PADDING ) == 0 )
+        {
+            uxSizeBytes += ipBUFFER_PADDING;
+
+            uint8_t * const pucBuffer = pucGetNetworkBuffer( &( uxSizeBytes ) );
+
+            if( pucBuffer == NULL )
+            {
+                pxNetworkBufferCopy = NULL;
+            }
+            else
+            {
+                pxNetworkBufferCopy->xDataLength = uxSizeBytes;
+
+                const size_t xOriginalLength = pxNetworkBufferCopy->xDataLength + ipBUFFER_PADDING;
+
+                if( uxSizeBytes > xOriginalLength )
+                {
+                    uxSizeBytes = xOriginalLength;
+                }
+
+                ( void ) memcpy( pucBuffer - ipBUFFER_PADDING,
+                                 pxNetworkBufferCopy->pucEthernetBuffer - ipBUFFER_PADDING,
+                                 uxSizeBytes );
+                vReleaseNetworkBuffer( pxNetworkBufferCopy->pucEthernetBuffer );
+                pxNetworkBufferCopy->pucEthernetBuffer = pucBuffer;
+            }
+        }
+        else
+        {
+            pxNetworkBufferCopy = NULL;
+        }
+
+        return pxNetworkBufferCopy;
+    }
+
+#endif /* if ipconfigIS_DISABLED( ipconfigBUFFER_ALLOC_STATIC ) */
+
+/*-----------------------------------------------------------*/
+
+#if ipconfigIS_ENABLED( ipconfigBUFFER_ALLOC_STATIC )
+
+    __attribute__( ( weak ) ) void vNetworkBufferAllocateRAMToBuffers( NetworkBufferDescriptor_t pxNetworkBuffers[ ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS ] )
```

Comment:
I think its better to keep the usage of compiler/target specific extensions such `weak`, `section`, etc.  as off this file to help not prevent build on compilers that doesnt support those syntax/extenstions.

Reply:
@tony-josi-aws, would you propose a macro like:
~~~
#define ipconfigWEAK_FUNCTION __attribute__( ( weak ) )

ipconfigWEAK_FUNCTION void vNetworkBufferAllocateRAMToBuffers(
                            NetworkBufferDescriptor_t pxNetworkBuffers[] )
~~~
?

In the kernel, there is a `portWEAK_SYMBOL`, but it is hardly used, only when using IAR.

Or make the declaration conditional:
~~~
#if( ipconfigINCLUDE_ALLOCATE_RAM_TO_BUFFERS_FUNCTION
    void vNetworkBufferAllocateRAMToBuffers(
          NetworkBufferDescriptor_t pxNetworkBuffers[] )
    {
    }
#endif
~~~

Reply:
Having macros might require us to add macros for all the extentions like  `weak`, `section`, `aligned`, etc thats used in the function and some compilers might follow entirely different syntax which might make it harder to maintain these macros.

I think second option is better (making the declaration conditional) or the old way of keeping the definition in the network interface file.

Reply:
I prefer the old way of keeping it in network interface file, is there any strong reason to move it to bufferallocation.c?

Reply:
Yes you're right, we also use `section` and `aligned`.

I have never heard users complain about the `__attribute__ ( ( aligned( 32 ) ) )` yet. Possibly people change the code themselves.

The same about `__attribute__( ( section( ".first_data" ) )` in the SAM project. `.first_data` is allocated in the first section of RAM, which is non-cacheable.

Reply:
The idea about making the definition conditional using a config value like `ipconfigINCLUDE_ALLOCATE_RAM_TO_BUFFERS_FUNCTION` was my original thought as well. If the kernel doesn't get updated to better support `portWEAK_SYMBOL` then perhaps the conditional is the better solution.